packages/browser-wallet/src/popup/store/transactions.ts

```diff
@@ -43,28 +44,51 @@ const monitorTransactionStatus = (genesisHash: string) => {
 
 const pendingTransactionsAtom = (() => {
     const base = atomWithChromeStorage<string[]>(ChromeStorageKey.PendingTransactions, []);
-
-    return atom<BrowserWalletAccountTransaction[], BrowserWalletAccountTransaction[]>(
+    const parsed = atom<BrowserWalletAccountTransaction[], BrowserWalletAccountTransaction[]>(
         (get) => {
-            const pendingJson = get(base);
-            const pending: BrowserWalletAccountTransaction[] = pendingJson.map(parse);
-            const network = get(networkConfigurationAtom);
+            const pending: BrowserWalletAccountTransaction[] = get(base).map(parse);
+            return pending;
+        },
+        (_, set, update) => {
+            set(base, update.map(stringify));
+        }
+    );
+
+    const derived = atom<BrowserWalletAccountTransaction[], BrowserWalletAccountTransaction[]>(
+        (get) => get(parsed),
+        (get, set, update) => {
+            const pending = [...get(parsed), ...update];
+            set(parsed, pending);
 
+            const network = get(networkConfigurationAtom);
             const monitor = monitorTransactionStatus(network.genesisHash);
+
             pending.forEach(async ({ transactionHash }) => {
                 const shouldRemove = await monitor(transactionHash);
+                const currentNetwork = get(networkConfigurationAtom);
 
-                if (shouldRemove) {
-                    sessionPendingTransactions.set(pendingJson.filter((json) => !json.includes(transactionHash)));
+                if (!shouldRemove) {
+                    return;
                 }
-            });
 
-            return pending;
-        },
-        (_, set, update) => {
-            set(base, update.map(stringify));
+                const networkChanged = network.genesisHash !== currentNetwork.genesisHash;
+                const nextPending = pending.filter((p) => p.transactionHash !== transactionHash);
+
+                if (!networkChanged) {
+                    set(parsed, nextPending);
+                } else {
+                    const spt = useIndexedStorage(sessionPendingTransactions, async () => currentNetwork.genesisHash);
+                    spt.set(nextPending.map(stringify));
```

Comment:
It seems weird that you do theses updates inside the forEach, instead of filtering and then doing a single update?

Reply:
I've done it this way because I want the transactions to be removed as soon as they've been verified as finalized. As I read it, what you propose is to wait for all pending transactions to be finalized, and then remove them in one single update? Or am I misunderstanding?

Reply:
> The monitor() blocks until it is finalized?

Correct.

> I guess the only issue is that pending is not updated, so the updates will cancel each other out?

I don't understand? Why is pending not updated? That's what happens in the 5 lines you marked?